reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package net.consensys.linea;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+import net.consensys.linea.zktracer.json.JsonConverter;
+
+@Slf4j
+public class MapFailedReferenceTestsTool {
+
+  public static void mapAndStoreFailedReferenceTest(
+      String testName, List<String> logEventMessages, String jsonOutputPath) {
+    Set<String> failedConstraints = getFailedConstraints(logEventMessages);
+    String jsonString = readFailedTestsOutput(jsonOutputPath);
+    JsonConverter jsonConverter = JsonConverter.builder().build();
+
+    List<ModuleToConstraints> constraintToFailingTests =
+        getModuleToConstraints(jsonString, jsonConverter);
+
+    mapFailedConstraintsToTestsToModule(failedConstraints, constraintToFailingTests, testName);
+
+    jsonString = jsonConverter.toJson(constraintToFailingTests);
+    writeToJsonFile(jsonString, jsonOutputPath);
+  }
+
+  private static void mapFailedConstraintsToTestsToModule(
+      Set<String> failedConstraints,
```

Comment:
are you creating here the different parts of the generated JSON? The name of the method does not look self-explanatory. I suggest to consider adding some comments or splitting the method into smaller ones.

Reply:
Yes this is the main method for mapping the failed tests, constraint and its parent module together.
I've split the method into smaller ones, so hopefully the method names help describe what's going on at each stage better now.